faculty/clients/template.py

```diff
@@ -42,4 +45,310 @@ def publish_new(self, project_id, template, source_directory):
             "sourceDirectory": source_directory,
             "name": template,
         }
-        return self._post_raw(endpoint, json=payload)
+        response = self._post_raw(endpoint, json=payload, check_status=False)
+        return _handle_publishing_response(response)
+
+    def add_to_project_from_directory(
+        self,
+        source_project_id,
+        source_directory,
+        target_project_id,
+        target_directory,
+        parameters,
+    ):
+        endpoint = "project/{}/apply".format(target_project_id)
+        payload = {
+            "sourceProjectId": str(source_project_id),
+            "sourceDirectory": source_directory,
+            "targetDirectory": target_directory,
+            "parameterValues": parameters,
+        }
+        response = self._post_raw(endpoint, json=payload, check_status=False)
+        return _handle_publishing_response(response)
+
+
+def _handle_publishing_response(response):
+    if 200 <= response.status_code < 300:
+        return response
+    elif 400 <= response.status_code < 500:
+        response_body = response.json()
+        error_code = response_body.get("errorCode")
+        if error_code == "resources_validation_failure":
+            raise ResourceValidationFailureResponseSchema().load(response_body)
+        elif error_code == "parameter_validation_failure":
+            raise ParameterValidationFailureSchema().load(response_body)
+        elif error_code == "template_retrieval_failure":
+            raise TemplateRetrievalFailureResponseSchema().load(response_body)
+        elif error_code == "default_parameters_parsing_error":
+            raise DefaultParametersParsingErrorSchema().load(response_body)
+        elif error_code == "generic_parsing_failure":
+            raise GenericParsingErrorSchema().load(response_body)
+        elif error_code == "workspace_files_validation_error":
+            raise WorkspaceFilesValidationErrorSchema().load(response_body)
+    else:
```

Comment:
There's the ServiceUnavailable response too, which this would swallow. Not sure we care but it's not an "Unexpected response".

Reply:
Yeah, I wasn't sure it's worth keeping this response in the backend and was planning to remove it (it's actually not present in the `publish new` flow. I'm not convinced it's worth the hassle to propagate this error if the server cant' process the request (compared to just throwing and IllegalStateException in the server and returning a 500).